src/query/sql/src/async_function/async_function_mgr.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use databend_common_ast::ast::Expr;
+use databend_common_base::base::GlobalInstance;
+use databend_common_catalog::catalog::Catalog;
+use databend_common_exception::ErrorCode;
+use databend_common_exception::Result;
+use databend_common_exception::Span;
+use databend_common_expression::types::DataType;
+use databend_common_meta_app::tenant::Tenant;
+
+use super::sequence_async_function::SequenceAsyncFunction;
+use crate::plans::AsyncFunctionCall;
+use crate::ScalarExpr;
+
+#[async_trait::async_trait]
+pub trait AsyncFunction: Sync + Send {
+    fn func_name(&self) -> String;
+
+    async fn generate(
```

Comment:
Too many args for the trait.

The `tenant`, `catalog` should be inside the struct which implements AsyncFunction